Zend/zend_weakrefs.c

```diff
@@ -0,0 +1,219 @@
+   +----------------------------------------------------------------------+
+*/
+
+/* $Id$ */
+
+#include "zend.h"
+#include "zend_interfaces.h"
+#include "zend_objects_API.h"
+#include "zend_weakrefs.h"
+
+typedef struct _zend_weakref {
+	zval referent;
+	zend_object std;
+} zend_weakref;
+
+zend_class_entry *zend_ce_weakref;
+zend_object_handlers zend_weakref_handlers;
+
+#define zend_weakref_from(o) ((zend_weakref*)(((char*) o) - XtOffsetOf(zend_weakref, std)))
+#define zend_weakref_fetch(z) zend_weakref_from(Z_OBJ_P(z))
+
+static void zend_weakrefs_dtor(zval *zv) {
+	zend_hash_destroy(Z_PTR_P(zv));
+	FREE_HASHTABLE(Z_PTR_P(zv));
+}
+
+void zend_weakrefs_init(HashTable *root) {
+	zend_hash_init(root, 8, NULL, zend_weakrefs_dtor, 0);
+}
+
+void zend_weakrefs_shutdown(HashTable *root) {
+	zend_hash_destroy(root);
+}
+
+static zend_object* zend_weakref_new(zend_class_entry *ce) {
+	zend_weakref *wr =
+		(zend_weakref*) 
+			zend_object_alloc(sizeof(zend_weakref), ce);
+
+	zend_object_std_init(&wr->std, ce);
+
+	wr->std.handlers = &zend_weakref_handlers;
+
+	ZVAL_NULL(&wr->referent);
+
+	return &wr->std;
+}
+
+static zend_object* zend_weakref_clone(zval *zv) {
+	zval rv;
+	zend_weakref *wr = zend_weakref_fetch(zv);
+
+	object_init_ex(&rv, Z_OBJCE_P(zv));
+
+	zend_weakref_construct(&rv, &wr->referent);
+
+	return Z_OBJ(rv);
+}
+
+static void zend_weakref_unref(zval *zv) {
+	zend_weakref *wr = 
+		zend_weakref_fetch(zv);
+
+	ZVAL_NULL(&wr->referent);
+
+	zval_ptr_dtor(zv);
+}
+
+static inline zend_array* zend_weakrefs_fetch(zend_object *ro) {
+	HashTable *refs = zend_hash_index_find_ptr(&EG(weakrefs), (zend_ulong) ro);
+
+	if (!refs) {
+		ALLOC_HASHTABLE(refs);
+		zend_hash_init(refs, 8, NULL, zend_weakref_unref, 0);
+		zend_hash_index_update_ptr(&EG(weakrefs), (zend_ulong) ro, refs);
+	}
+
+	return refs;
+}
+
+void zend_weakref_construct(zval *weakref, zval *referent) {
+	zend_weakref *wr = zend_weakref_fetch(weakref);
+
+	if (!(Z_OBJCE_P(referent)->ce_flags & ZEND_ACC_USE_WEAK_REFS)) {
+		zend_throw_error(NULL, 
+			"cannot weakly reference objects of type %s", 
+			ZSTR_VAL(Z_OBJCE_P(referent)->name));
+		return;
+	}
+
+	zend_hash_next_index_insert(
+		zend_weakrefs_fetch(Z_OBJ_P(referent)), weakref);
+
+	Z_ADDREF_P(weakref);
+
+	ZVAL_COPY_VALUE(&wr->referent, referent);
+
+	GC_ADD_FLAGS(Z_OBJ_P(referent), IS_OBJ_WEAKLY_REFERENCED);
+}
+
+#define zend_weakref_unsupported(thing) \
+	zend_throw_error(NULL, "weakref objects do not support " thing);
+
+static void zend_weakref_no_write(zval *object, zval *member, zval *value, void **rtc) {
+	zend_weakref_unsupported("properties");
+}
+
+static zval* zend_weakref_no_read(zval *object, zval *member, int type, void **rtc, zval *rv) {
+	if (!EG(exception)) {
+		zend_weakref_unsupported("properties");
+	}
+
+	return &EG(uninitialized_zval);
+}
+
+static zval *zend_weakref_no_read_ptr(zval *object, zval *member, int type, void **rtc) {
+	zend_weakref_unsupported("references");
```

Comment:
I think this should also just use `"properties"`. Or maybe `"property references"` if you want to be precise.